### DISCUSSION TITLE
Is it possible to use a Zod schema with a z.number in it as the base schema?

### DISCUSSION THREAD

**Discussion by zayenz** [airjp73/rvf#63] [Q&A] [answered]
I used some code that included a `z.number()` in the Zod schema. In the process of using the Zod schema with `zod-form-data`, I  tried to use the [suggested method](https://www.remix-validated-form.io/zod-form-data/api-reference#formData) to wrap a standard Zod schema using `zfd.formData`, but this didn't wrap the contained `z.number` in a `zfd.numeric` and thus the parsing failed.

Do I need to wrap all my numbers in my base schema in `zfd.numeric`, or can this be done automatically some way?

**Comment by airjp73** [comment, marked as answer]
Hi!

The number schema itself has to be wrapped with `zfd.numeric`. Most of the helpers are implemented with `z.preprocess` and don't alter your schema beyond that. `zfd.formData` only preprocesses the FormData into a plain object and handles nested array & object syntax (you don't actually need to use it if you're using with `remix-validated-form`).

**Reply by airjp73** [reply]
To provide a little more context, not doing this automatically is an intentional design decision. I want this library to work with all the zod apis without any extra limitations. The big one that causes issues with automatic transformations is `z.union` (which I personally use a decent amount). If you take a look at this (admittedly contrived) example, which transformation should we apply automatically to the `id` field? Because it could be multiple things, it's not really possible for us to say.

```ts
const schema = zfd.formData(
  z
    .union([
      z.object({
        subaction: z.literal("edit-person"),
        id: z.number()
      }),
      z.object({
        subaction: z.literal("edit-event"),
        id: z.string()
      })
    ])
    .and(
      z.object({
        name: z.string()
      })
    )
);
```

I'm also aware that there are other tools (like remix-params-helper) that do process this automatically, but [that has to handle each zod type explicitly](https://github.com/kiliman/remix-params-helper/blob/0f202aeac8c0f75855b1cf9298dae1de470c68a1/src/helper.ts#L190-L222) and I don't think it has support for `ZodUnion`, `ZodEffects`, or `ZodIntersection`.

So that's why I opted for this design rather than automatic processing. It's a little bit more noise in the schema, but it's going to be a lot more reliable and maintainable in the long run.

**Reply by zayenz** [reply]
Thanks for the detailed reply, and I get the trade-offs. Good information to know.
